File: codebase/src-cpp/test/ng6/object/RegisterObjectTest.cpp
/*
 *   Copyright 2008 The Portico Project
 *
 *   This file is part of portico.
 *
 *   portico is free software; you can redistribute it and/or modify
 *   it under the terms of the Common Developer and Distribution License (CDDL)
 *   as published by Sun Microsystems. For more information see the LICENSE file.
 *
 *   Use of this software is strictly AT YOUR OWN RISK!!!
 *   If something bad happens you do not have permission to come crying to me.
 *   (that goes for your lawyer as well)
 *
 */
#include "RegisterObjectTest.h"

// Register test suite with the global repository
CPPUNIT_TEST_SUITE_REGISTRATION( RegisterObjectTest );
CPPUNIT_TEST_SUITE_NAMED_REGISTRATION( RegisterObjectTest, "RegisterObjectTest" );
CPPUNIT_TEST_SUITE_NAMED_REGISTRATION( RegisterObjectTest, "registerObject" );
CPPUNIT_TEST_SUITE_NAMED_REGISTRATION( RegisterObjectTest, "objectManagement" );

/////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////// Constructors/Destructors /////////////////////////////////
/////////////////////////////////////////////////////////////////////////////////////////////
RegisterObjectTest::RegisterObjectTest()
{
	this->defaultFederate = new TestNG6Federate( "defaultFederate" );
	this->federateA = new TestNG6Federate( "federateA" );
	this->federateB = new TestNG6Federate( "federateB" );
	this->federateNone = new TestNG6Federate( "federateNone" );
}

RegisterObjectTest::~RegisterObjectTest()
{
	delete this->defaultFederate;
	delete this->federateA;
	delete this->federateB;
	delete this->federateNone;
}

/////////////////////////////////////////////////////////////////////////////////////////////
/////////////////////////////// Test Setup and Helper Methods ///////////////////////////////
/////////////////////////////////////////////////////////////////////////////////////////////

void RegisterObjectTest::setUp()
{
    this->defaultFederate->quickCreate();
    this->defaultFederate->quickJoin();
    this->federateA->quickJoin();
    this->federateB->quickJoin();
    this->federateNone->quickJoin();
    
    // get the handle information
    this->aHandle  = defaultFederate->quickOCHandle( "ObjectRoot.A" );
    this->bHandle  = defaultFederate->quickOCHandle( "ObjectRoot.A.B" );
    
    // publish and subscribe
    this->defaultFederate->quickPublish( "ObjectRoot.A.B", 1, "aa" );
    this->federateA->quickSubscribe( "ObjectRoot.A", 1, "aa" );
    this->federateB->quickSubscribe( "ObjectRoot.A.B", 1, "aa" );
}

void RegisterObjectTest::tearDown()
{
	this->federateNone->quickResign();
	this->federateB->quickResign();
	this->federateA->quickResign();
	this->defaultFederate->quickResign();
	this->defaultFederate->quickDestroy();
}

////////////////////////////////////////////////////////////////////////////////////////////
/////////////////////////////////////// Test Methods ///////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////
// TEST: (valid) testRegisterObject() //
////////////////////////////////////////
void RegisterObjectTest::testRegisterObject()
{
	RTI::ObjectHandle theObject = 0;
	try
	{
		theObject = defaultFederate->rtiamb->registerObjectInstance( bHandle );
	}
	catch( RTI::Exception& e )
	{
		failTest( "Unexpected exception while registering object instance: %s", e._reason );
	}
	
	// make sure it is discovered by the other federates
	federateA->fedamb->waitForDiscoveryAs( theObject, aHandle );
	federateB->fedamb->waitForDiscoveryAs( theObject, bHandle );
	federateNone->fedamb->waitForDiscoveryTimeout( theObject );
}

//////////////////////////////////////////////////////
// TEST: testRegisterObjectWithInvalidClassHandle() //
//////////////////////////////////////////////////////
void RegisterObjectTest::testRegisterObjectWithInvalidClassHandle()
{
	try
	{
		defaultFederate->rtiamb->registerObjectInstance( (RTI::ObjectClassHandle)100000 );
		failTestMissingException( "ObjectClassNotDefined",
		                          "registering instance with invalid class handle" );
	}
	catch( RTI::ObjectClassNotDefined& ocnd )
	{
		// success!
	}
	catch( RTI::Exception& e )
	{
		failTestWrongException( "ObjectClassNotDefined", e,
		                        "registering instance with invalid class handle" );
	}
}

///////////////////////////////////////////////////////////
// TEST: testRegisterObjectWithNonPublishedClassHandle() //
///////////////////////////////////////////////////////////
void RegisterObjectTest::testRegisterObjectWithNonPublishedClassHandle()
{
	try
	{
		defaultFederate->rtiamb->registerObjectInstance( aHandle );
		failTestMissingException( "ObjectClassNotPublished",
		                          "registering instance with unpublished class handle" );
	}
	catch( RTI::ObjectClassNotPublished& ocnp )
	{
		// success!
	}
	catch( RTI::Exception& e )
	{
		failTestWrongException( "ObjectClassNotPublished", e,
		                        "registering instance with unpublished class handle" );
	}
}

////////////////////////////////////////////////////////////////////////////////////////////
/////////////////////////////// Register Named Object Tests ////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////

/////////////////////////////////////////////
// TEST: (valid) testRegisterNamedObject() //
/////////////////////////////////////////////
void RegisterObjectTest::testRegisterNamedObject()
{
	RTI::ObjectHandle theObject = 0;
	try
	{
		theObject = defaultFederate->rtiamb->registerObjectInstance( bHandle, "MyObject" );
	}
	catch( RTI::Exception& e )
	{
		failTest( "Unexpected exception while registering object instance: %s", e._reason );
	}
	
	// make sure it is discovered by the other federates
	federateA->fedamb->waitForDiscoveryAsWithName( theObject, aHandle, "MyObject" );
	federateB->fedamb->waitForDiscoveryAsWithName( theObject, bHandle, "MyObject" );
	federateNone->fedamb->waitForDiscoveryTimeout( theObject );
}

/////////////////////////////////////////////////////////
// TEST: (valid) testRegisterNamedObjectWithNullName() //
/////////////////////////////////////////////////////////
void RegisterObjectTest::testRegisterNamedObjectWithNullName()
{
	RTI::ObjectHandle theObject = 0;
	try
	{
		theObject = defaultFederate->rtiamb->registerObjectInstance( bHandle, NULL );
	}
	catch( RTI::Exception& e )
	{
		failTest( "Unexpected exception while registering object instance: %s", e._reason );
	}
	
	// make sure it is discovered by the other federates
	char *expectedName = new char[8];
	sprintf( expectedName, "HLA%lo", theObject );
	federateA->fedamb->waitForDiscoveryAsWithName( theObject, aHandle, expectedName );
	federateB->fedamb->waitForDiscoveryAsWithName( theObject, bHandle, expectedName );
	delete expectedName; // the waitForDiscoveryAsWithName() will delete if the test fails
	federateNone->fedamb->waitForDiscoveryTimeout( theObject );
}

///////////////////////////////////////////////////////////////
// TEST: (valid) testRegisterNamedObjectWithWhitespaceName() //
///////////////////////////////////////////////////////////////
void RegisterObjectTest::testRegisterNamedObjectWithWhitespaceName()
{
	RTI::ObjectHandle theObject = 0;
	try
	{
		theObject = defaultFederate->rtiamb->registerObjectInstance( bHandle, "   " );
	}
	catch( RTI::Exception& e )
	{
		failTest( "Unexpected exception while registering object instance: %s", e._reason );
	}
	
	// make sure it is discovered by the other federates
	char *expectedName = new char[4];
	sprintf( expectedName, "   " );
	federateA->fedamb->waitForDiscoveryAsWithName( theObject, aHandle, expectedName );
	federateB->fedamb->waitForDiscoveryAsWithName( theObject, bHandle, expectedName );
	delete expectedName; // the waitForDiscoveryAsWithName() will delete if the test fails
	federateNone->fedamb->waitForDiscoveryTimeout( theObject );
}

///////////////////////////////////////////////////////////
// TEST: testRegisterNamedObjectWithInvalidClassHandle() //
///////////////////////////////////////////////////////////
void RegisterObjectTest::testRegisterNamedObjectWithInvalidClassHandle()
{
	try
	{
		defaultFederate->rtiamb->registerObjectInstance( (RTI::ObjectClassHandle)100000, "MyName" );
		failTestMissingException( "ObjectClassNotDefined",
		                          "registering instance with invalid class handle" );
	}
	catch( RTI::ObjectClassNotDefined& ocnd )
	{
		// success!
	}
	catch( RTI::Exception& e )
	{
		failTestWrongException( "ObjectClassNotDefined", e,
		                        "registering instance with invalid class handle" );
	}
}

/////////////////////////////////////////////////////
// TEST: testRegisterNamedObjectWithExistingName() //
/////////////////////////////////////////////////////
void RegisterObjectTest::testRegisterNamedObjectWithExistingName()
{
	RTI::ObjectClassHandle firstObject = defaultFederate->quickRegister( bHandle, "MyObject" );
	federateA->fedamb->waitForDiscoveryAsWithName( firstObject, aHandle, "MyObject" );
	federateB->fedamb->waitForDiscoveryAsWithName( firstObject, bHandle, "MyObject" );
	
	try
	{
		defaultFederate->rtiamb->registerObjectInstance( bHandle, "MyObject" );
		failTestMissingException( "ObjectAlreadyRegistered",
		                          "registering instance with name that has already been taken" );
	}
	catch( RTI::ObjectAlreadyRegistered& ocnd )
	{
		// success!
	}
	catch( RTI::Exception& e )
	{
		failTestWrongException( "ObjectAlreadyRegistered", e,
		                        "registering instance with name that has already been taken" );
	}
}

////////////////////////////////////////////////////////////////
// TEST: testRegisterNamedObjectWithNonPublishedClassHandle() //
////////////////////////////////////////////////////////////////
void RegisterObjectTest::testRegisterNamedObjectWithNonPublishedClassHandle()
{
	try
	{
		defaultFederate->rtiamb->registerObjectInstance( aHandle, "MyObject" );
		failTestMissingException( "ObjectClassNotPublished",
		                          "registering instance with unpublished class handle" );
	}
	catch( RTI::ObjectClassNotPublished& ocnp )
	{
		// success!
	}
	catch( RTI::Exception& e )
	{
		failTestWrongException( "ObjectClassNotPublished", e,
		                        "registering instance with unpublished class handle" );
	}
}

File: codebase/src/java/test/hlaunit/hla13/object/RegisterObjectTest.java
/*
 *   Copyright 2006 The Portico Project
 *
 *   This file is part of portico.
 *
 *   portico is free software; you can redistribute it and/or modify
 *   it under the terms of the Common Developer and Distribution License (CDDL) 
 *   as published by Sun Microsystems. For more information see the LICENSE file.
 *   
 *   Use of this software is strictly AT YOUR OWN RISK!!!
 *   If something bad happens you do not have permission to come crying to me.
 *   (that goes for your lawyer as well)
 *
 */
package hlaunit.hla13.object;

import hla.rti.FederateNotExecutionMember;
import hla.rti.ObjectAlreadyRegistered;
import hla.rti.ObjectClassNotDefined;
import hla.rti.ObjectClassNotPublished;
import hla.rti.RestoreInProgress;
import hla.rti.SaveInProgress;
import hlaunit.hla13.common.Abstract13Test;
import hlaunit.hla13.common.Test13Federate;
import hlaunit.hla13.common.Test13Instance;

import org.portico.lrc.PorticoConstants;
import org.testng.Assert;
import org.testng.annotations.AfterClass;
import org.testng.annotations.AfterMethod;
import org.testng.annotations.BeforeClass;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

@Test(singleThreaded=true, groups={"RegisterObjectTest", "registerObject", "objectManagement"})
public class RegisterObjectTest extends Abstract13Test
{
	//----------------------------------------------------------
	//                    STATIC VARIABLES
	//----------------------------------------------------------

	//----------------------------------------------------------
	//                   INSTANCE VARIABLES
	//----------------------------------------------------------
	private Test13Federate secondFederate;
	private int aHandle, aaHandle, abHandle, acHandle;

	//----------------------------------------------------------
	//                      CONSTRUCTORS
	//----------------------------------------------------------

	//----------------------------------------------------------
	//                    INSTANCE METHODS
	//----------------------------------------------------------
	@BeforeClass(alwaysRun=true)
	public void beforeClass()
	{
		super.beforeClass();

		secondFederate = new Test13Federate( "secondFederate", this );
	}
	
	@BeforeMethod(alwaysRun=true)
	public void beforeMethod()
	{
		defaultFederate.quickCreate();
		defaultFederate.quickJoin();
		secondFederate.quickJoin();
		
		// cache the handles
		aHandle  = defaultFederate.quickOCHandle( "ObjectRoot.A" );
		aaHandle = defaultFederate.quickACHandle( "ObjectRoot.A", "aa" );
		abHandle = defaultFederate.quickACHandle( "ObjectRoot.A", "ab" );
		acHandle = defaultFederate.quickACHandle( "ObjectRoot.A", "ac" );
		
		// do publication and subscription
		defaultFederate.quickPublish( "ObjectRoot.A", "aa", "ab" );
		secondFederate.quickSubscribe( "ObjectRoot.A", "aa", "ab", "ac" );
	}
	
	@AfterMethod(alwaysRun=true)
	public void afterMethod()
	{
		secondFederate.quickResign();
		defaultFederate.quickResign();
		defaultFederate.quickDestroy();
	}
	
	@AfterClass(alwaysRun=true)
	public void afterClass()
	{
		super.afterClass();
	}
	
	//////////////////////////////////////////////////////////////////////////////////////////
	////////////////////////////////////// Test Methods //////////////////////////////////////
	//////////////////////////////////////////////////////////////////////////////////////////
	// int registerObjectInstance( int classHandle )
	//     throws ObjectClassNotDefined,
	//            ObjectClassNotPublished,
	//            FederateNotExecutionMember,
	//            SaveInProgress,
	//            RestoreInProgress,
	//            RTIinternalError,
	//            ConcurrentAccessAttempted;
	
	////////////////////////////////////////
	// TEST: (valid) testRegisterObject() //
	////////////////////////////////////////
	@Test
	public void testRegisterObject()
	{
		// register an object instance
		int oHandle = -1;
		try
		{
			oHandle = defaultFederate.rtiamb.registerObjectInstance( aHandle );
		}
		catch( Exception e )
		{
			Assert.fail( "Exception while registering object instance", e );
		}
		
		// make sure the other federates discover it
		secondFederate.fedamb.waitForDiscovery( oHandle );
	}

	///////////////////////////////////////////////////////////////////
	// TEST: testRegisterObjectFederateOnlyOwnsPublishedAttributes() //
	///////////////////////////////////////////////////////////////////
	@Test
	public void testRegisterObjectFederateOnlyOwnsPublishedAttributes()
	{
		// register an object instance
		int oHandle = -1;
		try
		{
			oHandle = defaultFederate.rtiamb.registerObjectInstance( aHandle );
		}
		catch( Exception e )
		{
			Assert.fail( "Exception while registering object instance", e );
		}
		
		// make sure the other federates discover it
		secondFederate.fedamb.waitForDiscovery( oHandle );
		
		// validate that the correct ownership is in place
		try
		{
			boolean aa = defaultFederate.rtiamb.isAttributeOwnedByFederate( oHandle, aaHandle );
			boolean ab = defaultFederate.rtiamb.isAttributeOwnedByFederate( oHandle, abHandle );
			boolean ac = defaultFederate.rtiamb.isAttributeOwnedByFederate( oHandle, acHandle );
			
			Assert.assertTrue( aa, "Federate doesn't own attribute aa, it should" );
			Assert.assertTrue( ab, "Federate doesn't own attribute ab, it should" );
			Assert.assertFalse( ac, "Federate owns attribute ac, it should not" );
		}
		catch( Exception e )
		{
			Assert.fail( "Unexpected exception determining ownership of attributes" , e );
		}
	}

	//////////////////////////////////////////////////////
	// TEST: testRegisterObjectWithInvalidClassHandle() //
	//////////////////////////////////////////////////////
	@Test
	public void testRegisterObjectWithInvalidClassHandle()
	{
		try
		{
			defaultFederate.rtiamb.registerObjectInstance( 111111 );
			expectedException( ObjectClassNotDefined.class );
		}
		catch( ObjectClassNotDefined ocnd )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, ObjectClassNotDefined.class );
		}
	}
	
	///////////////////////////////////////////////////////////
	// TEST: testRegisterObjectWithNonPublishedClassHandle() //
	///////////////////////////////////////////////////////////
	@Test
	public void testRegisterObjectWithNonPublishedClassHandle()
	{
		try
		{
			// the second federate does not publish anything
			secondFederate.rtiamb.registerObjectInstance( aHandle );
			expectedException( ObjectClassNotPublished.class );
		}
		catch( ObjectClassNotPublished ocnp )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, ObjectClassNotPublished.class );
		}
	}

	/////////////////////////////////////////////
	// TEST: testRegisterObjectWhenNotJoined() //
	/////////////////////////////////////////////
	@Test
	public void testRegisterObjectWhenNotJoined()
	{
		defaultFederate.quickResign();
		try
		{
			defaultFederate.rtiamb.registerObjectInstance( aHandle );
			expectedException( FederateNotExecutionMember.class );
		}
		catch( FederateNotExecutionMember fnem )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, FederateNotExecutionMember.class );
		}
	}

	//////////////////////////////////////////////////
	// TEST: testRegisterObjectWhenSaveInProgress() //
	//////////////////////////////////////////////////
	@Test
	public void testRegisterObjectWhenSaveInProgress()
	{
		defaultFederate.quickSaveInProgress( "save" );
		try
		{
			defaultFederate.rtiamb.registerObjectInstance( aHandle );
			expectedException( SaveInProgress.class );
		}
		catch( SaveInProgress sip )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, SaveInProgress.class );
		}
	}

	/////////////////////////////////////////////////////
	// TEST: testRegisterObjectWhenRestoreInProgress() //
	/////////////////////////////////////////////////////
	@Test
	public void testRegisterObjectWhenRestoreInProgress()
	{
		defaultFederate.quickRestoreInProgress( "save" );
		try
		{
			defaultFederate.rtiamb.registerObjectInstance( aHandle );
			expectedException( RestoreInProgress.class );
		}
		catch( RestoreInProgress rip )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, RestoreInProgress.class );
		}
	}

	//////////////////////////////////////////////////////////////////////////////////////////
	//////////////////////////////// Object Name Test Methods ////////////////////////////////
	//////////////////////////////////////////////////////////////////////////////////////////
	// int registerObjectInstance( int classHandle, String name )
	//     throws ObjectClassNotDefined,
	//            ObjectClassNotPublished,
	//            ObjectAlreadyRegistered,
	//            FederateNotExecutionMember,
	//            SaveInProgress,
	//            RestoreInProgress,
	//            RTIinternalError,
	//            ConcurrentAccessAttempted;

	/////////////////////////////////////////////
	// TEST: (valid) testRegisterNamedObject() //
	/////////////////////////////////////////////
	@Test
	public void testRegisterNamedObject()
	{
		// register an object instance
		int oHandle = -1;
		String oName = "myobject";
		try
		{
			oHandle = defaultFederate.rtiamb.registerObjectInstance( aHandle, oName );
		}
		catch( Exception e )
		{
			Assert.fail( "Exception while registering named object instance", e );
		}
		
		// make sure the other federates discover it
		secondFederate.fedamb.waitForDiscovery( oHandle );
		Test13Instance instance = secondFederate.fedamb.getInstances().get( oHandle );
		Assert.assertEquals( instance.getName(), oName,
		                     "Discovered Object didn't have the correct name" );
	}

	/////////////////////////////////////////////////////////
	// TEST: (valid) testRegisterNamedObjectWithNullName() //
	/////////////////////////////////////////////////////////
	/**
	 * This should be allowed, but the registered object should have an RTI generated name.
	 * Null should not be used for it.
	 */
	@Test
	public void testRegisterNamedObjectWithNullName()
	{
		int oHandle = -1;
		try
		{
			oHandle = defaultFederate.rtiamb.registerObjectInstance( aHandle, null );
		}
		catch( Exception e )
		{
			unexpectedException( "registering object instance with null name", e );
		}
		
		// verify that the object was registered
		secondFederate.fedamb.waitForDiscovery( oHandle );
		Test13Instance instance = secondFederate.fedamb.getInstances().get( oHandle );
		Assert.assertNotNull( instance.getName(), "Discovered object has null name" );
	}

	///////////////////////////////////////////////////////////////
	// TEST: (valid) testRegisterNamedObjectWithWhitespaceName() //
	///////////////////////////////////////////////////////////////
	/**
	 * As with registering an object with a null name, this should also work. However, the
	 * whitespace name used should be given to the instance. Although it isn't very elegant
	 * to use such a value for a name, it shouldn't be illegal
	 */
	@Test
	public void testRegisterNamedObjectWithWhitespaceName()
	{
		int oHandle = -1;
		String name = "   ";
		try
		{
			oHandle = defaultFederate.rtiamb.registerObjectInstance( aHandle, name );
		}
		catch( Exception e )
		{
			unexpectedException( "registering object instance with whitespace name", e );
		}
		
		// verify that the object was registered
		secondFederate.fedamb.waitForDiscovery( oHandle );
		Test13Instance instance = secondFederate.fedamb.getInstances().get( oHandle );
		Assert.assertEquals( instance.getName(), name,
		                     "Discovered object should have whitespace name" );
	}

	///////////////////////////////////////////////////////////
	// TEST: testRegisterNamedObjectWithInvalidClassHandle() //
	///////////////////////////////////////////////////////////
	@Test
	public void testRegisterNamedObjectWithInvalidClassHandle()
	{
		try
		{
			defaultFederate.rtiamb.registerObjectInstance( 111111, "myobject" );
			expectedException( ObjectClassNotDefined.class );
		}
		catch( ObjectClassNotDefined ocnd )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, ObjectClassNotDefined.class );
		}
	}

	/////////////////////////////////////////////////////
	// TEST: testRegisterNamedObjectWithExistingName() //
	/////////////////////////////////////////////////////
	@Test
	public void testRegisterNamedObjectWithExistingName()
	{
		// register the first object
		int handleOne = -1;
		String nameOne = "nameOne";
		
		try
		{
			handleOne = defaultFederate.rtiamb.registerObjectInstance( aHandle, nameOne );
		}
		catch( Exception e )
		{
			Assert.fail( "Exception while registering named object instance", e );
		}
		
		// wait for the object to be discovered
		secondFederate.fedamb.waitForDiscovery( handleOne );
		
		// try and register another object with the same name
		try
		{
			defaultFederate.rtiamb.registerObjectInstance( aHandle, nameOne );
			expectedException( ObjectAlreadyRegistered.class );
		}
		catch( ObjectAlreadyRegistered oar )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, ObjectAlreadyRegistered.class );
		}
	}
	
	//////////////////////////////////////////////////////////////////////////
	// TEST: testRegisterNamedObjectWithExistingNameFromDifferentFederate() //
	//////////////////////////////////////////////////////////////////////////
	@Test
	public void testRegisterNamedObjectWithExistingNameFromDifferentFederate()
	{
		int objectHandle = defaultFederate.quickRegister( "ObjectRoot.A", "myObject" );
		secondFederate.fedamb.waitForDiscovery( objectHandle );
		
		// make sure the second federate is publishing as required
		secondFederate.quickPublish( "ObjectRoot.A", "aa" );
		
		try
		{
			secondFederate.rtiamb.registerObjectInstance( aHandle, "myObject" );
			expectedException( ObjectAlreadyRegistered.class );
		}
		catch( ObjectAlreadyRegistered oar )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, ObjectAlreadyRegistered.class );
		}
	}
	
	///////////////////////////////////////////////////////////////
	// TEST: testRegisterNamedObjectWithExistingNameAtSameTime() //
	///////////////////////////////////////////////////////////////
	@Test
	public void testRegisterNamedObjectWithExistingNameAtSameTime()
	{
		// make sure that object name checking is ENABLED!
		System.setProperty( PorticoConstants.PROPERTY_NEGOTIATE_OBJECT_NAMES, "true" );
		
		try
		{
        	secondFederate.quickPublish( "ObjectRoot.A", "aa" );
        	
        	// register the object in the default federate, but don't tick the second federate,
        	// so it shouldn't get the discovery notificiation
        	defaultFederate.quickRegister( "ObjectRoot.A", "myObject" );
        	
        	try
        	{
        		secondFederate.rtiamb.registerObjectInstance( aHandle, "myObject" );
        		expectedException( ObjectAlreadyRegistered.class );
        	}
        	catch( ObjectAlreadyRegistered oar )
        	{
        		// success!
        	}
        	catch( Exception e )
        	{
        		wrongException( e, ObjectAlreadyRegistered.class );
        	}
		}
		finally
		{
			// back to the default value
			System.setProperty( PorticoConstants.PROPERTY_NEGOTIATE_OBJECT_NAMES, "false" );
		}
	}
	
	////////////////////////////////////////////////////////////////
	// TEST: testRegisterNamedObjectWithNonPublishedClassHandle() //
	////////////////////////////////////////////////////////////////
	@Test
	public void testRegisterNamedObjectWithNonPublishedClassHandle()
	{
		try
		{
			// the second federate does not publish anything
			secondFederate.rtiamb.registerObjectInstance( aHandle, "myobject" );
			expectedException( ObjectClassNotPublished.class );
		}
		catch( ObjectClassNotPublished ocnp )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, ObjectClassNotPublished.class );
		}
	}
	
	////////////////////////////////////////////////////////////////////////
	// TEST: testRegisterNamedObjectFederateOnlyOwnsPublishedAttributes() //
	////////////////////////////////////////////////////////////////////////
	@Test
	public void testRegisterNamedObjectFederateOnlyOwnsPublishedAttributes()
	{
		// register an object instance
		int oHandle = -1;
		try
		{
			oHandle = defaultFederate.rtiamb.registerObjectInstance( aHandle, "myobject" );
		}
		catch( Exception e )
		{
			Assert.fail( "Exception while registering named object instance", e );
		}
		
		// make sure the other federates discover it
		secondFederate.fedamb.waitForDiscovery( oHandle );
		Test13Instance instance = secondFederate.fedamb.getInstances().get( oHandle );
		Assert.assertEquals( instance.getName(), "myobject",
		                     "Discovered Object didn't have the correct name" );

		// validate that the correct ownership is in place
		try
		{
			boolean aa = defaultFederate.rtiamb.isAttributeOwnedByFederate( oHandle, aaHandle );
			boolean ab = defaultFederate.rtiamb.isAttributeOwnedByFederate( oHandle, abHandle );
			boolean ac = defaultFederate.rtiamb.isAttributeOwnedByFederate( oHandle, acHandle );
			
			Assert.assertTrue( aa, "Federate doesn't own attribute aa, it should" );
			Assert.assertTrue( ab, "Federate doesn't own attribute ab, it should" );
			Assert.assertFalse( ac, "Federate owns attribute ac, it should not" );
		}
		catch( Exception e )
		{
			Assert.fail( "Unexpected exception determining ownership of attributes" , e );
		}
	}

	//////////////////////////////////////////////////
	// TEST: testRegisterNamedObjectWhenNotJoined() //
	//////////////////////////////////////////////////
	@Test
	public void testRegisterNamedObjectWhenNotJoined()
	{
		defaultFederate.quickResign();
		try
		{
			defaultFederate.rtiamb.registerObjectInstance( aHandle, "myObject" );
			expectedException( FederateNotExecutionMember.class );
		}
		catch( FederateNotExecutionMember fnem )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, FederateNotExecutionMember.class );
		}
	}

	///////////////////////////////////////////////////////
	// TEST: testRegisterNamedObjectWhenSaveInProgress() //
	///////////////////////////////////////////////////////
	@Test
	public void testRegisterNamedObjectWhenSaveInProgress()
	{
		defaultFederate.quickSaveInProgress( "save" );
		try
		{
			defaultFederate.rtiamb.registerObjectInstance( aHandle, "myObject" );
			expectedException( SaveInProgress.class );
		}
		catch( SaveInProgress sip )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, SaveInProgress.class );
		}
	}

	//////////////////////////////////////////////////////////
	// TEST: testRegisterNamedObjectWhenRestoreInProgress() //
	//////////////////////////////////////////////////////////
	@Test
	public void testRegisterNamedObjectWhenRestoreInProgress()
	{
		defaultFederate.quickRestoreInProgress( "save" );
		try
		{
			defaultFederate.rtiamb.registerObjectInstance( aHandle, "myObject" );
			expectedException( RestoreInProgress.class );
		}
		catch( RestoreInProgress rip )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, RestoreInProgress.class );
		}
	}

	//----------------------------------------------------------
	//                     STATIC METHODS
	//----------------------------------------------------------
}
